File: source/c/lista_enlazada.c
// ==========================================
// ESTRUCTURA DE DATOS: LISTA ENLAZADA
// ==========================================
// Explicación Detallada:
// Una lista enlazada permite almacenar datos de forma dinámica en la memoria.
// A diferencia de un arreglo estático, no necesitamos conocer la cantidad de elementos de antemano.
// Cada "Nodo" se aloja en memoria dinámica (usando malloc) y se enlaza con el siguiente
// mediante el puntero 'siguiente'.

#include <stdio.h>
#include <stdlib.h>

// Definición de la estructura del Nodo
struct Nodo {
    int dato;               // Almacena el valor numérico
    struct Nodo* siguiente; // Puntero al siguiente nodo en la cadena
};

// INSERCIÓN (al final)
// Proceso:
// 1. Asignamos memoria para el nuevo nodo con 'malloc'.
// 2. Inicializamos su 'dato' y hacemos que 'siguiente' apunte a NULL (ya que será el último).
// 3. Si la lista está vacía (*cabeza == NULL), el nuevo nodo será la cabeza.
// 4. Si ya tiene elementos, usamos un puntero auxiliar ('ultimo') para recorrer la lista
//    hasta llegar al final, y luego enlazamos el último nodo con el nuevo.
void insertar(struct Nodo** cabeza, int nuevo_dato) {
    // 1. Crear nuevo nodo
    struct Nodo* nuevo_nodo = (struct Nodo*)malloc(sizeof(struct Nodo));
    struct Nodo* ultimo = *cabeza; // Puntero auxiliar
    
    // 2. Inicializar valores
    nuevo_nodo->dato = nuevo_dato;
    nuevo_nodo->siguiente = NULL;
    
    // 3. Si está vacía, hacer del nuevo nodo la cabeza
    if (*cabeza == NULL) {
        *cabeza = nuevo_nodo;
        return;
    }
    
    // 4. Recorrer hasta el final y enlazar
    while (ultimo->siguiente != NULL) {
        ultimo = ultimo->siguiente;
    }
    ultimo->siguiente = nuevo_nodo;
}

// RECORRIDO
// Proceso:
// Empezamos desde el primer nodo, e imprimimos su valor.
// Luego avanzamos al nodo siguiente iterativamente hasta que el puntero sea NULL.
void recorrer(struct Nodo* nodo) {
    printf("Lista Enlazada: ");
    while (nodo != NULL) {
        printf("%d -> ", nodo->dato);
        nodo = nodo->siguiente;
    }
    printf("NULL\n");
}

// BÚSQUEDA
// Proceso:
// Compara secuencialmente el valor de cada nodo con el 'dato_buscado'.
// Devuelve 1 (Verdadero) si lo encuentra, o 0 (Falso) si llega al final.
int buscar(struct Nodo* nodo, int dato_buscado) {
    while (nodo != NULL) {
        if (nodo->dato == dato_buscado) return 1;
        nodo = nodo->siguiente;
    }
    return 0;
}

int main() {
    struct Nodo* cabeza = NULL; // La lista inicia vacía
    insertar(&cabeza, 10);
    insertar(&cabeza, 20);
    insertar(&cabeza, 30);
    recorrer(cabeza);
    printf("¿El 20 está en la lista?: %s\n", buscar(cabeza, 20) ? "Sí" : "No");
    return 0;
}

File: source/python/lista_enlazada.py
# ==========================================
# ESTRUCTURA DE DATOS: LISTA ENLAZADA
# ==========================================
# Explicación Detallada:
# Una lista enlazada es una estructura de datos lineal que consiste en una 
# secuencia de elementos (nodos). A diferencia de los arrays (arreglos convencionales), 
# los elementos de una lista enlazada no están almacenados en ubicaciones de memoria contiguas. 
# En cambio, cada nodo contiene un enlace (referencia o puntero) al siguiente nodo de la secuencia.
# Ventajas: Tamaño dinámico, inserción/eliminación eficiente sin desplazar elementos.

class Nodo:
    # La clase Nodo representa cada elemento individual de la lista.
    def __init__(self, dato):
        self.dato = dato         # 'dato' almacena la información útil (número, texto, etc.)
        self.siguiente = None    # 'siguiente' es la referencia al próximo nodo (None si es el último)

class ListaEnlazada:
    def __init__(self):
        # Al crear la lista, inicialmente no hay elementos, por lo que la 'cabeza' apunta a None.
        self.cabeza = None

    # INSERCIÓN (al final de la lista)
    # Proceso:
    # 1. Se crea un nuevo nodo con el dato proporcionado.
    # 2. Si la lista está vacía (self.cabeza es None), el nuevo nodo se convierte en el primer elemento.
    # 3. Si no está vacía, iteramos nodo por nodo usando un bucle 'while' hasta llegar al último nodo.
    # 4. Enlazamos el puntero 'siguiente' del último nodo al nuevo nodo que acabamos de crear.
    def insertar(self, dato):
        nuevo_nodo = Nodo(dato)
        if not self.cabeza:
            self.cabeza = nuevo_nodo
            return
        actual = self.cabeza
        while actual.siguiente:      # Avanzar mientras haya un nodo siguiente
            actual = actual.siguiente
        actual.siguiente = nuevo_nodo # El último nodo ahora apunta al nuevo nodo

    # RECORRIDO
    # Proceso:
    # 1. Empezamos en la cabeza (el primer elemento).
    # 2. Mientras el nodo actual no sea None, leemos su dato.
    # 3. Actualizamos la referencia al siguiente nodo (actual = actual.siguiente).
    def recorrer(self):
        elementos = []
        actual = self.cabeza
        while actual:
            elementos.append(actual.dato)
            actual = actual.siguiente
        print("Lista Enlazada:", " -> ".join(map(str, elementos)))

    # BÚSQUEDA
    # Proceso:
    # 1. Recorremos secuencialmente los nodos comparando el 'dato_buscado' con el 'dato' del nodo actual.
    # 2. Si coinciden, retornamos True (elemento encontrado).
    # 3. Si llegamos al final de la lista sin coincidencias, retornamos False.
    def buscar(self, dato_buscado):
        actual = self.cabeza
        while actual:
            if actual.dato == dato_buscado:
                return True
            actual = actual.siguiente
        return False

if __name__ == "__main__":
    lista = ListaEnlazada()
    lista.insertar(10)
    lista.insertar(20)
    lista.insertar(30)
    lista.recorrer()
    print("¿El 20 está en la lista?:", lista.buscar(20))